nbresuse/prometheus.py

```diff
@@ -61,3 +72,52 @@ def apply_cpu_limit(self, cpu_metric_values) -> Optional[float]:
                 return self.config.cpu_limit
             else:
                 return 100.0 * cpu_metric_values["cpu_count"]
+
+    @gen.coroutine
+    def kernel_metrics(self):
+        kernels = yield maybe_future(self._list_kernel_memory())
+        for kernel_id, data in kernels.items():
+            self.KERNEL_MEMORY_USAGE.labels(kernel_id=kernel_id).set(data['rss'])
+
+    @gen.coroutine
+    def _list_kernel_memory(self):
+        kernel_memory = dict()
+
+        sessions = yield maybe_future(self.session_manager.list_sessions())
+        specs = yield maybe_future(self.kernel_spec_manager.get_all_specs())
+        kernel_processes = PrometheusHandler._kernel_processes(specs)
+
+        def find_process(kernel_id):
+            for proc in kernel_processes:
+                cmd = proc.cmdline()
+                if cmd:
+                    last_arg = cmd[-1]
+                    if kernel_id in last_arg:
+                        return proc
+            return None
+
+        for session in sessions:
+            kernel = session['kernel']
+            kernel_id = kernel['id']
+            kernel_process = find_process(kernel_id)
+            if kernel_process:
+                kernel_memory[kernel_id] = kernel_process.memory_info()._asdict()
+
+        raise gen.Return(kernel_memory)
+
+    @staticmethod
+    def _kernel_processes(kernel_specs):
+        for proc in process_iter():
+            try:
+                for specs in kernel_specs.values():
+                    for spec in specs.values():
+                        if 'argv' in spec:
+                            if spec.get('argv')[0] in {'python', 
+                                                       'python%i' % sys.version_info[0], 
+                                                       'python%i.%i' % sys.version_info[:2]}:
+                                spec.get('argv')[0] = sys.executable
+                                key = " ".join(spec.get('argv')[:-1])
+                                if key in " ".join(proc.cmdline()):
+                                    yield proc
```

Comment:
Shouldn't that block be move one indentation to the left, to also include other kernels that are not python?

Maybe it could be useful to refactor this logic to reduce the level of nesting?

Reply:
Let me look into how I can reduce the level of nesting.

Reply:
Please check now.